openedx/features/enterprise_support/api.py

```diff
@@ -592,109 +593,148 @@ def enterprise_customer_for_request(request):
+    """
+    for record in existing_dsc_records:
+        if record['consent_provided']:
+            record_consent_data = (record['username'], str(record['enterprise_customer_uuid']), record['course_id'])
+            if record_consent_data == learner_consent_input_data:
+                return True
+    return False
+
+
 @enterprise_is_enabled(otherwise=False)
 def consent_needed_for_course(request, user, course_id, enrollment_exists=False):
     """
-    Wrap the enterprise app check to determine if the user needs to grant
+    Wraps the enterprise app check to determine if the user needs to grant
     data sharing permissions before accessing a course.
     """
     LOGGER.info(
-        "Determining if user [{username}] must consent to data sharing for course [{course_id}]".format(
+        "[ENTERPRISE DSC] Determining if user [{username}] must consent to data sharing for course"
+        " [{course_id}]".format(
             username=user.username,
             course_id=course_id
         )
     )
 
-    enterprise_learner_details = get_enterprise_learner_data_from_db(user)
-    enterprise_customer_uuid = None
-    if enterprise_learner_details:
-        enterprise_customer = enterprise_learner_details[0]['enterprise_customer']
-        enterprise_customer_uuid = enterprise_customer['uuid']
+    active_enterprise_customer_user = get_active_enterprise_customer_user(user)
+    if not active_enterprise_customer_user:
+        # user is not linked to any enterprise so return False
+        LOGGER.info(
+            "[ENTERPRISE DSC] Consent from user [{username}] is not needed for course [{course_id}]."
+            " The user is not linked to an enterprise.".format(
+                username=user.username,
+                course_id=course_id
+            )
+        )
+        return False
+
+    enterprise_customer = active_enterprise_customer_user['enterprise_customer']
 
-    consent_cache_key = get_data_consent_share_cache_key(user.id, course_id, enterprise_customer_uuid)
+    consent_cache_key = get_data_consent_share_cache_key(user.id, course_id, enterprise_customer['uuid'])
     data_sharing_consent_needed_cache = TieredCache.get_cached_response(consent_cache_key)
     if data_sharing_consent_needed_cache.is_found and data_sharing_consent_needed_cache.value == 0:
         LOGGER.info(
-            "Consent from user [{username}] is not needed for course [{course_id}]. The DSC cache was checked,"
-            " and the value was 0.".format(
+            "[ENTERPRISE DSC] Consent from user [{username}] is not needed for course [{course_id}]. "
+            "The DSC cache was checked and the value was 0.".format(
                 username=user.username,
                 course_id=course_id
             )
         )
         return False
 
-    consent_needed = False
-    if not enterprise_learner_details:
+    enable_data_sharing_consent = enterprise_customer['enable_data_sharing_consent']
```

Comment:
nit: not sure whether we should check whether customer have enabled DSC or Cache first...

Reply:
In my opinion, it's always best to check the cache first before doing any sort of processing. That's why the check for cache is placed on top